### PR TITLE
Support CMYK color

### LINKED ISSUE #747: Can't get black value of a pixel in a CMYK Image
ImageMagick 6.9.10-49
I tried ruby 2.5.1 and ruby 2.6.1
I tried RMagick 3.2.0, 3.1.0 and 2.16.0

I try to read the black value of some pixels in a CMYK image without luck.

I always get back `0`

I checked ImageMagick with this command:

```bash
$ convert black.jpg -format '%[pixel:p{1,1}]' info:-
cmyk(0,0,0,255)%
```

When I try to get the same value with RMagick with this script

```ruby
require 'rmagick'

black = Magick::Image.read('black.jpg')[0]
black_pixels = []
black.each_pixel do |pixel|
    black_pixels << pixel
end
puts "black.jpg Pixel 1/1: cmyk(#{black_pixels.first.cyan},#{black_pixels.first.magenta},#{black_pixels.first.yellow},#{black_pixels.first.black}),"

cyan = Magick::Image.read('cyan.jpg')[0]
cyan_pixels = []
cyan.each_pixel do |pixel|
    cyan_pixels << pixel
end
puts "cyan.jpg Pixel 1/1: cmyk(#{cyan_pixels.first.cyan},#{cyan_pixels.first.magenta},#{cyan_pixels.first.yellow},#{cyan_pixels.first.black})"
```

I get these results:

```bash
ruby read_color.rb
black.jpg Pixel 1/1: cmyk(0,0,0,0),
cyan.jpg Pixel 1/1: cmyk(65278,0,0,0)
```
Other color values are working, but not the black value.

I suspect that the error has something to do with the fact that black is associated with 'opacity' in ext/RMagick/rmpixel.c 

```c
DEF_PIXEL_CMYK_CHANNEL_ACCESSOR(cyan, red)
DEF_PIXEL_CMYK_CHANNEL_ACCESSOR(magenta, green)
DEF_PIXEL_CMYK_CHANNEL_ACCESSOR(yellow, blue)
DEF_PIXEL_CMYK_CHANNEL_ACCESSOR(black, opacity)
```
Here is my demo code: [https://github.com/onlinetocode/rmagick_black_test](https://github.com/onlinetocode/rmagick_black_test)

### PR DESCRIPTION
**Pull request by Watson1978**
Fix #747

I think `Image#pixel_color` also needs to fix

### CODE REVIEW

File: ext/RMagick/rmimage.c [hunk 1]
```diff
@@ -7113,29 +7113,33 @@ Image_get_pixels(VALUE self, VALUE x_arg, VALUE y_arg, VALUE cols_arg, VALUE row
...
-    if (!pixels)
+    if (columns * rows <= 0)
     {
         return rb_ary_new();
     }
 
     // Allocate an array big enough to contain the PixelPackets.
-    size = (long)(columns * rows);
-    pixel_ary = rb_ary_new2(size);
+    pixel_ary = rb_ary_new2(columns * rows);
```
Comment (Watson1978) [inline ext/RMagick/rmimage.c]: Hmm, I think `pixel_ary` just requires `(columns - x) * (rows - y)` pixel size. `columns * rows` is too much.

Comment (Watson1978) [reply ext/RMagick/rmimage.c]: I'm guessing the original `for (n = 0; n < size; n++)` is also wrong...
if `x` or `y` is not 0, `pixel_ary` will has unexpected value in trailing region....

**Comment by Watson1978** [comment]
Hmm, seems that `QuantumPixelPacket` is not supported in ImageMagick 6.7...

**Comment by onlinetocode** [comment]
I added a demo for changing the black value into my code ([onlinetocode/rmagick_black_test](https://github.com/onlinetocode/rmagick_black_test))

You could run it via `bundle exec ruby setting_black.rb`

Setting the black value via `Pixel#black=` works. But when I store the changed pixels back via `Image#store_pixels` the new black value is ignored and the initial black value is preserved. I hope these findings are a bit helpful.

**Comment by dlemstra** [comment]
It appears that I can push changes to your branch @Watson1978. I did not realize in the UI that I was pushing it to your branch. I added an extra commit that creates an extra struct that we can use for this situation. It now also works for the 6.7 version of ImageMagick.

**Pull request closed without merging**